### PR TITLE
Plonky3 + Goldilocks: Use Poseidon2 instead of Poseidon

### PR DESCRIPTION
**Pull request by georgwiese**
After discussion with the Polygon Zero team, this should be significantly faster when using AVX, while still recursion friendly.

Tested this by running the Poseidon example (on our server) using the following script:
```
cargo run -r --bin powdr-rs compile riscv/tests/riscv_data/keccak-o output --field bb
cargo run -r --features plonky3,halo2 pil output/keccak.asm -o output -f --field bb --prove-with plonky3-composite
```

I got these numbers:
```
# main (stable Rust)

./keccak.sh
3.0752053s

RUSTFLAGS='-C target-cpu=native' ./keccak.sh
5.150651s

(using nightly-features and nightly Rust)
RUSTFLAGS='-C target-cpu=native' ./keccak.sh
3.8711894s

# use-poseidon2 (stable Rust)

./keccak.sh
2.4612818s

RUSTFLAGS='-C target-cpu=native' ./keccak.sh
2.2226317s

(using nightly-features and nightly Rust)
RUSTFLAGS='-C target-cpu=native' ./keccak.sh
1.8559729s
```

If I understand this right, this is still not using AVX, because we have stable Rust and have not activated `nightly-features`. But the AVX implementation of Poseidon on Goldilocks used a placeholder implementation that was slower than the non-AVX implementation, so the Rust flag actually harmed. This seems to be fixed by switching to Poseidon2.

### CODE REVIEW

File: plonky3/src/params/baby_bear.rs [hunk 1]
```diff
@@ -20,10 +20,6 @@ use rand::{distributions::Standard, Rng, SeedableRng};
 use powdr_number::BabyBearField;
 
 const D: u64 = 7;
-// params directly taken from plonky3's poseidon2_round_numbers_128 function
-// to guarantee 128-bit security.
-const ROUNDS_F: usize = 8;
-const ROUNDS_P: usize = 13;
```
Comment (georgwiese) [inline plonky3/src/params/baby_bear.rs:26]: Didn't really see a reason to hard-code this here. If for any reason the `poseidon2_round_numbers_128`  function changes, we'll want to use the updated numbers.

File: plonky3/src/params/baby_bear.rs [hunk 2]
```diff
@@ -58,17 +54,20 @@ const FRI_PROOF_OF_WORK_BITS: usize = 16;
 const RNG_SEED: u64 = 42;
 
 lazy_static! {
+    static ref ROUNDS: (usize, usize) = poseidon2_round_numbers_128::<Goldilocks>(WIDTH, D);
```
Comment (georgwiese) [inline plonky3/src/params/baby_bear.rs]: Seems like `static ref (ROUNDS_F, ROUNDS_P): (usize, usize)` is not possible, not sure if there's a better way!

File: plonky3/src/params/baby_bear.rs (hunk 1 above)
Comment (georgwiese) [reply plonky3/src/params/baby_bear.rs:26]: @Schaeff LMK if you had a reason I'm not aware of :)

**Review by leonardoalt — DISMISSED** [review]

**Comment by leonardoalt** [comment]
has conflicts now

**Comment by leonardoalt** [comment]
with https://github.com/powdr-labs/powdr/pull/1869 which is the head of the queue

File: plonky3/src/params/baby_bear.rs (hunk 1 above)
Comment (Schaeff) [reply plonky3/src/params/baby_bear.rs:26]: I don't think there is a reason for it, I did not write this code but missed that in the review.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**